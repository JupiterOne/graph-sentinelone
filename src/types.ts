import { IntegrationInstanceConfig } from '@jupiterone/integration-sdk-core';

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  serverUrl: string;
  apiToken: string;
}
