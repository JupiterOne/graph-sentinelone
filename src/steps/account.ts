import {
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { IntegrationConfig } from '../types';
import { Entities } from './constants';
import { createAccountEntity } from './converters';

export const ACCOUNT_ENTITY_KEY = 'entity:account';

export async function fetchAccountDetails({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const accountEntity = await jobState.addEntity(createAccountEntity(instance));
  await jobState.setData(ACCOUNT_ENTITY_KEY, accountEntity);
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-account',
    name: 'Fetch Account Details',
    entities: [Entities.ACCOUNT],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccountDetails,
  },
];
