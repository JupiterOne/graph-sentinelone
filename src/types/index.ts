import { EntityFromIntegration } from "@jupiterone/jupiter-managed-integration-sdk";

export type AccountEntity = EntityFromIntegration;

export interface GroupEntity extends EntityFromIntegration {
  inherits?: boolean;
  name?: string;
  creator?: string;
  filterName?: string;
  totalAgents?: number;
  filterId?: string;
  rank?: number;
  siteId: string;
  isDefault?: boolean;
  creatorId?: string;
  updatedAt?: number;
  type?: string;
  id: string;
  createdAt?: number;
}

export interface AgentEntity extends EntityFromIntegration {
  domain?: string;
  appsVulnerabilityStatus?: string;
  siteName?: string;
  coreCount?: number;
  totalMemory?: number;
  inRemoteShellSession?: boolean;
  osArch?: string;
  allowRemoteShell?: boolean;
  scanStatus?: string;
  consoleMigrationStatus?: string;
  updatedAt?: number;
  osType?: string;
  id: string;
  createdAt?: number;
  externalIp?: string;
  computerName?: string;
  modelName?: string;
  uuid?: string;
  encryptedApplications?: boolean;
  adComputerDistinguishedName?: string;
  osUsername?: string;
  groupName?: string;
  infected?: boolean;
  policyUpdatedAt?: number;
  cpuId?: string;
  registeredAt?: number;
  activeThreats?: number;
  groupUpdatedAt?: number;
  machineType?: string;
  groupIp?: string;
  osStartTime?: number;
  osRevision?: string;
  scanAbortedAt?: number;
  siteId?: string;
  scanStartedAt?: number;
  isPendingUninstall?: boolean;
  scanFinishedAt?: number;
  lastActiveDate?: number;
  groupId: string;
  isActive?: boolean;
  agentVersion?: string;
  licenseKey?: string;
  networkStatus?: string;
  lastLoggedInUserName?: string;
  osName?: string;
  mitigationMode?: string;
  cpuCount?: number;
  isUninstalled?: boolean;
  isUpToDate?: boolean;
  mitigationModeSuspicious?: string;
  isDecommissioned?: boolean;
}
