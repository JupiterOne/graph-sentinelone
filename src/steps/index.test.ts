import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';

import { IntegrationConfig } from '../types';
import { fetchAccountDetails } from './account';

// import { fetchAgents, fetchGroups } from './agents';

const DEFAULT_SERVER_URL = 'https://usea1-partners.sentinelone.net';
const DEFAULT_API_TOKEN = 'dummy-api-token';

const integrationConfig: IntegrationConfig = {
  serverUrl: process.env.SERVER_URL || DEFAULT_SERVER_URL,
  apiToken: process.env.API_TOKEN || DEFAULT_API_TOKEN,
};

test('should collect data', async () => {
  const context = createMockStepExecutionContext<IntegrationConfig>({
    instanceConfig: integrationConfig,
  });

  // Simulates dependency graph execution.
  // See https://github.com/JupiterOne/sdk/issues/262.
  await fetchAccountDetails(context);
  // await fetchAgents(context);
  // await fetchGroups(context);

  // Review snapshot, failure is a regression
  expect({
    numCollectedEntities: context.jobState.collectedEntities.length,
    numCollectedRelationships: context.jobState.collectedRelationships.length,
    collectedEntities: context.jobState.collectedEntities,
    collectedRelationships: context.jobState.collectedRelationships,
    encounteredTypes: context.jobState.encounteredTypes,
  }).toMatchSnapshot();

  const accounts = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('Account'),
  );
  expect(accounts.length).toBeGreaterThan(0);
  expect(accounts).toMatchGraphObjectSchema({
    _class: ['Account'],
    schema: {
      additionalProperties: false,
      properties: {
        _type: { const: 'sentinelone_account' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
    },
  });

  // const users = context.jobState.collectedEntities.filter((e) =>
  //   e._class.includes('User'),
  // );
  // expect(users.length).toBeGreaterThan(0);
  // expect(users).toMatchGraphObjectSchema({
  //   _class: ['User'],
  //   schema: {
  //     additionalProperties: false,
  //     properties: {
  //       _type: { const: 'acme_user' },
  //       firstName: { type: 'string' },
  //       _rawData: {
  //         type: 'array',
  //         items: { type: 'object' },
  //       },
  //     },
  //     required: ['firstName'],
  //   },
  // });

  // const userGroups = context.jobState.collectedEntities.filter((e) =>
  //   e._class.includes('UserGroup'),
  // );
  // expect(userGroups.length).toBeGreaterThan(0);
  // expect(userGroups).toMatchGraphObjectSchema({
  //   _class: ['UserGroup'],
  //   schema: {
  //     additionalProperties: false,
  //     properties: {
  //       _type: { const: 'acme_group' },
  //       logoLink: {
  //         type: 'string',
  //         // Validate that the `logoLink` property has a URL format
  //         format: 'url',
  //       },
  //       _rawData: {
  //         type: 'array',
  //         items: { type: 'object' },
  //       },
  //     },
  //     required: ['logoLink'],
  //   },
  // });
});
