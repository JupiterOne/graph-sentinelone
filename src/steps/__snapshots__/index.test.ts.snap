// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should collect data 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Account",
      ],
      "_key": "sentinelone_account-local-integration-instance",
      "_rawData": [],
      "_type": "sentinelone_account",
      "createdOn": undefined,
      "displayName": "Local Integration",
      "name": "Local Integration",
      "vendor": "SentinelOne",
    },
  ],
  "collectedRelationships": [],
  "encounteredTypes": [
    "sentinelone_account",
  ],
  "numCollectedEntities": 1,
  "numCollectedRelationships": 0,
}
`;
